File: Python/Leetcode Easy/225ImplementStackUsingQueues.py
#https://leetcode.com/problems/implement-stack-using-queues/description/
class MyStack:
    def __init__(self):
        self.queue1 = []
        self.queue2 = []

    def push(self, x: int) -> None:
        if len(self.queue1) == 0:
            self.queue1.append(x)
            for item in self.queue2:
                self.queue1.append(item)
            self.queue2 = []
        else:
            self.queue2.append(x)
            for item in self.queue1:
                self.queue2.append(item)
            self.queue1 = []

    def pop(self) -> int:
        if len(self.queue1) == 0:
            return self.queue2.pop(0)
        else:
            return self.queue1.pop(0)

    def top(self) -> int:
        if len(self.queue1) == 0:
            return self.queue2[0]
        else:
            return self.queue1[0]
        

    def empty(self) -> bool:
        return len(self.queue1) == 0 and len(self.queue2) == 0
        


# Your MyStack object will be instantiated and called as such:
# obj = MyStack()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.top()
# param_4 = obj.empty()

File: C++/Leetcode Easy/225ImplementStackUsingQueues.cpp
#include <list>
using namespace std;
class MyStack {
public:
    list<int> q1;
    list<int> q2;
    MyStack() {
    }
    
    void push(int x) {
        if (q1.empty()) {
            q1.push_back(x);
            while (!q2.empty()) {
                q1.push_back(q2.front());
                q2.pop_front();
            }
        }
        else {
            q2.push_back(x);
            while (!q1.empty()) {
                q2.push_back(q1.front());
                q1.pop_front();
            }
        }
    }
    
    int pop() {
        int val = top();
        if (!q1.empty()) q1.pop_front();
        else q2.pop_front();
        return val;
    }
    
    int top() {
        if (!q1.empty()) return q1.front();
        else return q2.front();
    }
    
    bool empty() {
        return q1.empty() && q2.empty();
    }
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */